src/permutations.jl

```diff
@@ -34,42 +72,48 @@ Generate all size `t` permutations of an indexable object `a`.
 """
 function permutations(a, t::Integer)
     if t < 0
-        t = length(a) + 1
+        return []
```

Comment:
I would write it as:
```
t == 0 && [Vector{eltype(a)}()]
1 <= t =< length(a)) || Vector{eltype(a)}[]
```

Reply:
@mschauer suggests an empty `PermutationsIterator` (although I'm not really sure how to implement that). 

Is there a good reason to prefer one over the other?

Reply:
Returning `PermutationIterator` would make the code type stable.
I will propose changes that will fix this.

Reply:
in this function change the code to the old one:
```
if t < 0
    t = length(a) + 1
end
return PermutationIterator(a, t)
```
(I cannot do it as I do not have enough edit rights). The rest will be in an edit

Reply:
Based on the suggestion above, I re-wrote this function as follows.
```julia
function permutations(a, t::Integer)
    if t < 0
        t = length(a) + 1
    end
    return PermutationIterator(a, t)
end
```

However it fails the following tests:
```julia
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:17
  Test threw exception
  Expression: collect(permutations("abc", 4)) == Any[]
  DivideError: integer division error
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:21
  Test threw exception
  Expression: collect(permutations("abc", -1)) == Any[]
  DivideError: integer division error
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:22
  Test threw exception
  Expression: collect(permutations("", 1)) == Any[]
  DivideError: integer division error
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:24
  Test threw exception
  Expression: collect(permutations("", -1)) == Any[]
  DivideError: integer division error
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:29
  Test threw exception
  Expression: collect(permutations([], -1)) == Any[]
  DivideError: integer division error
permutations: Error During Test at /Users/natemcintosh/dev/Combinatorics.jl/test/permutations.jl:31
  Test threw exception
  Expression: collect(permutations([], 1)) == Any[]
  DivideError: integer division error
```
Digging into it a little more: 
`collect(permutations("", 1))` gets stuck on `_similar_shape(itr, isz)`, which throws
`Internal Error: DivideError: integer division error`

(`[p for p in permutations("", 1)]` hits the same error)

The call `_similar_shape(itr, isz)` in this case turns out to be `length(itr)`, where `itr` is `Base.Generator{PermutationIterator{String}, typeof(identity)}(identity, PermutationIterator{String}("", 1))`
I am unable to see inside the call to `length()`, for some reason the debugger just throws the exception, and I can't step into the call

I thought perhaps I can get around this issue by changing `t` to 0?
No. This causes test failures because Vector{Any}[[]] != Any[]
I don't understand the distinction between what we return in the case t==0, and the case
`t<0` and `t>length(a)`.
- In the case we do not care about the distinction, we could simply always set `t = 0`
when it is <0 or >length(a), and change the tests.
- In the case we do care, I am stuck on how to fix the above issue

Out of curiosity, I also tested the code below. It goes into the iterate method for this type, and correctly returns `nothing`
```julia
for p in permutations("", 1)
    @show p
end
```

Reply:
No, the error is elsewhere. You defined `Base.length` incorrectly I think. I have proposed a fix.

Reply:
Thank you!